### DISCUSSION TITLE
PorousFlow: how to visualise velocity on surfaces (paraview)

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#24132] [Q&A Tools] [answered]
Hi guys,

Our mesh has surfaces that we apply BCs on them, I want to visualise the velocity normal to the surfaces, is it possible with MOOSE/Paraview?

```
[./topol_linear_TOPOLSRF_F005_UG1]
type = PorousFlowPiecewiseLinearSinkBEH
variable = porepressure
boundary = 'TOPOLSRF_F005_UG1'
pt_vals = '-1e9 1e9'
multipliers = '-1e9 1e9'
flux_function = 1
v = min_dist
PT_shift = -9810
fluid_phase = 0
[../]

```
I know I can visualise the surface via Sets on Paraview, but I only can show node properties (velocity is a mesh property)

![image](https://user-images.githubusercontent.com/79562439/233235823-66dbe889-cc54-4846-823c-d04e452a02f3.png)

Is there a way to work around this (like project element velocity value to the element surface [normal multiply?])?

Thank you very much!

Kind regards,
Traiwit

**Comment by GiudGiud** [comment]
Hello

You ll want to look at auxkernels for that.
Create a new auxiliary variable vec_times_n, const monomial

Then in the auxkernel, also boundary restricted, you can set the value with velocity * _normal

You ll want the DIffusionFluxAux but change from grad_u to just _u. 

There might be one already somewhere, but cant seem to find it

Guillaume

**Reply by Traiwit** [reply]
Hi @GiudGiud,

Thank you for a reply, to start, should I create another object based on `DIffusionFluxAux` and follow what you mentioned? 

* vec_times_n, const monomial

* set the value with velocity * _normal

* change from grad_u to just _u.

Kind regards,
Traiwit

**Reply by GiudGiud** [reply]
I would do that yes. 

vec_times_n, const monomial -> just an input file change

New auxkernel (named: AdvectiveFluxAux)
- parameters: vx vy vz as params.addCoupledVar
- coupled variables v_xyz set as v_x(coupledValue("vx")
- return RealVectorValue(v_x[_qp], v_y[_qp], v_z[_qp]) * _normals[_qp]

**Reply by Traiwit** [reply]
this is my code:

```
#include "AdvectiveFluxAuxBE.h"
#include "Assembly.h"

registerMooseObject("MOOSE_BEHApp", AdvectiveFluxAuxBE);

InputParameters
AdvectiveFluxAuxBE::validParams()
{
  InputParameters params = AuxKernel::validParams();
  params.addClassDescription("Compute components of flux vector for advection problems "
                             "$(\\vec{J} = \\vec{v} \\cdot \\vec{n})$.");
  params.addCoupledVar("vx", "The x component of velocity");
  params.addCoupledVar("vy", "The y component of velocity");
  params.addCoupledVar("vz", "The z component of velocity");
  return params;
}

AdvectiveFluxAuxBE::AdvectiveFluxAuxBE(const InputParameters & parameters)
  : AuxKernel(parameters),
    _v_x(coupledValue("vx")),
    _v_y(coupledValue("vy")),
    _v_z(coupledValue("vz")),
    _normals(_assembly.normals())
{
}

Real
AdvectiveFluxAuxBE::computeValue()
{
  RealVectorValue velocity(_v_x[_qp], _v_y[_qp], _v_z[_qp]);
  return velocity * _normals[_qp];
}
```

i'm still a bit confused, where does it take the velocity data from?

I look at the example of the input file, it will look somewhat like this:

```

[AuxVariables]
  [./flux_n]
    order = CONSTANT
    family = MONOMIAL
  [../]
[]

[AuxKernels]
  [./flux_n]
    type = AdvectiveFluxAuxBE
    variable = flux_n
    component = normal
    boundary = 'xxxx'  #define my surface here
    check_boundary_restricted = false
  [../]
[]
```

and how do I visualise this data? 

Kind regards,
Traiwit

**Reply by GiudGiud** [reply]
> i'm still a bit confused, where does it take the velocity data from?

Do you not have velocity non linear variables?
And if not, cant porous flow compute them from the pressure field?

> and how do I visualise this data?

You visualize it on the whole domain, then use the threhshold filter in paraview to only show the nonzero cells. They will all be near the boundary

**Reply by Traiwit** [reply]
> Do you not have velocity non linear variables?
> And if not, cant porous flow compute them from the pressure field?

hmm i'm using this to compute Velocity `PorousFlowDarcyVelocityComponent`

```
[./darcy_x]
order = CONSTANT
family = MONOMIAL
[../]
[./darcy_y]
order = CONSTANT
family = MONOMIAL
[../]
[./darcy_z]
order = CONSTANT
family = MONOMIAL
[../]
[]


[AuxKernels]
[./darcy_x]
type = PorousFlowDarcyVelocityComponent
variable = darcy_x
gravity = '0 0 -9.81'
component = x
[../]
[./darcy_y]
type = PorousFlowDarcyVelocityComponent
variable = darcy_y
gravity = '0 0 -9.81'
component = y
[../]
[./darcy_z]
type = PorousFlowDarcyVelocityComponent
variable = darcy_z
gravity = '0 0 -9.81'
component = z
[../]
[]
```

I'm still not sure, I will get back to you once there's an error.

> You visualize it on the whole domain, then use the threhshold filter in paraview to only show the nonzero cells. They will all be near the boundary

This makes sense, thanks!

Kind regards,
Traiwit

**Reply by GiudGiud** [reply]
darcy_xyz could serve as the v_xyz parameters then I think.

btw if you feel like making a pull request this auxkernel could go in the moose/framework. We are keen on external contributions

**Reply by Traiwit** [reply, marked as answer]
@GiudGiud 

I think it kinda almost works now, this is the updated code:

```
#include "AdvectiveFluxAuxBE.h"
#include "Assembly.h"

registerMooseObject("MOOSE_BEHApp", AdvectiveFluxAuxBE);

InputParameters
AdvectiveFluxAuxBE::validParams()
{
  InputParameters params = AuxKernel::validParams();
  MooseEnum component("x y z normal");
  params.addClassDescription("Compute components of flux vector for advection problems "
                             "$(\\vec{J} = \\vec{v} \\cdot \\vec{n})$.");
  params.addCoupledVar("v", "The x component of velocity");
  params.addParam<MooseEnum>("component", component, "The desired component of flux."); // Added this line
  return params;
}

AdvectiveFluxAuxBE::AdvectiveFluxAuxBE(const InputParameters & parameters)
  : AuxKernel(parameters),
    _component(getParam<MooseEnum>("component")),
    _v(coupledValue("v")),
    _normals(_assembly.normals()),
    _use_normal(getParam<MooseEnum>("component") == "normal")
{
  if (_use_normal && !isParamValid("boundary"))
    paramError("boundary", "A boundary must be provided if using the normal component!");
}

Real
AdvectiveFluxAuxBE::computeValue()
{
  RealVectorValue velocity(_v[_qp]);
  return velocity * _normals[_qp];
}
```

in this particular case, I only do it for flux_x

```
  [./flux_x]
    type = AdvectiveFluxAuxBE
    variable = flux_x
    v = darcy_x
    component = normal
    boundary = DRAINSRF_F008_UG4
    check_boundary_restricted = false
  [../]

```

just wondering why in the Darcy_x it shows negative value (which is correct based on the flow and drainage conditions) but when I project it to the surface (figure below), it becomes positive, is there anything that I missed?

Kind regards,
Traiwit 
![image](https://user-images.githubusercontent.com/79562439/235045273-3aa4928b-91ec-4026-b583-d41671f7d64a.png)

**Reply by GiudGiud** [reply]
How is the sideset oriented? The _normal attribute will depend on that

**Reply by Traiwit** [reply]
> How is the sideset oriented? The _normal attribute will depend on that

I think it kinda makes sense now, it positive becase the velocity is flowing outward respect to the surface (hence positive) not to the xyz space. on the other hands if the flow is inward into the surface, it will be positive (i guess)

![image](https://user-images.githubusercontent.com/79562439/235049793-11e05627-bdeb-46e6-8278-fda12eab30d6.png)

above is the surface where I apply BCs

```
[./drain_DRAINSRF_F008_UG4] 
type = PorousFlowPiecewiseLinearSinkBEH 
variable = porepressure 
boundary = 'DRAINSRF_F008_UG4'
pt_vals = '0 1e9' 
multipliers = '0 1e9' 
flux_function = 1
v = min
PT_shift = 0 
use_mobility = true
fluid_phase = 0
[../]
```

**Reply by GiudGiud** [reply]
> I think it kinda makes sense now, it positive becase the velocity is flowing outward respect to the surface (hence positive) not to the xyz space. on the other hands if the flow is inward into the surface, it will be positive (i guess)

yes, as long as the sideset is consistently oriented

how is that surface/sideset defined? using a commercial mesher?
Does it consistently point the same direction?

We unfortunately do not currently have a way (coming this summer is the plan) to look at sideset directions from MOOSE

**Reply by Traiwit** [reply]
> yes, as long as the sideset is consistently oriented
> 
> how is that surface/sideset defined? using a commercial mesher? Does it consistently point the same direction?

Yes i think so, from Abaqus.

> We unfortunately do not currently have a way (coming this summer is the plan) to look at sideset directions from MOOSE

This will be awesome! looking forward to have this feature soon!

all good for now, thank you @GiudGiud!

**Reply by Traiwit** [reply]
Hi @GiudGiud,

Apparently we kinda want the nx ny and nz to compute the scalar quantity we want
![image](https://user-images.githubusercontent.com/79562439/236141086-0b1b7708-086d-4735-8f89-d6475c494609.png)

not sure what this `_normals[_qp]` actually does in the code?

```
    _normals(_assembly.normals()),
    _use_normal(getParam<MooseEnum>("component") == "normal")
{
  if (_use_normal && !isParamValid("boundary"))
    paramError("boundary", "A boundary must be provided if using the normal component!");
}

Real
AdvectiveFluxAuxBE::computeValue()
{
  RealVectorValue velocity(_v[_qp]);
  return velocity * _normals[_qp];
```

is there a way that we can obtain this data from the surface? (we can externally compute it but it will be another set of .CSV which is not always ideal for our large simulations)


I will tag @WilkAndy here too, since he is a hydro expert (might be good to add flux across surface to PorousFlow)

Kind regards,
Traiwit

**Reply by WilkAndy** [reply]
I agree it'd be cool to "add flux across a surface to PorousFlow".  We've actually got that ability, but it's a bit painful to do in practice.  You use a https://mooseframework.inl.gov/source/bcs/PorousFlowOutflowBC.html applied to a boundary, and use a `save_in` to record the flow through each finite-element node.  But this will alter your boundary conditions, so you need to define another boundary that is identical to the first, but with its normal oriented the other way, and apply another https://mooseframework.inl.gov/source/bcs/PorousFlowOutflowBC.html to that one.  Then the BCs cancel out, but you've recorded one of them into a `save_in`.

**Reply by Traiwit** [reply]
> I agree it'd be cool to "add flux across a surface to PorousFlow". We've actually got that ability, but it's a bit painful to do in practice. You use a https://mooseframework.inl.gov/source/bcs/PorousFlowOutflowBC.html applied to a boundary, and use a `save_in` to record the flow through each finite-element node. But this will alter your boundary conditions, so you need to define another boundary that is identical to the first, but with its normal oriented the other way, and apply another https://mooseframework.inl.gov/source/bcs/PorousFlowOutflowBC.html to that one. Then the BCs cancel out, but you've recorded one of them into a `save_in`.

Hi @WilkAndy 

I dont get how to "define another boundary that is identical to the first, but with its normal oriented the other way" ?

I guess set `multiplier = -1` ?

so the input will be like:

```
[BCs]
  [outflow]
    type = PorousFlowOutflowBC
    boundary = 'DRAINSRF_F009_UG5'
    variable = pp
    save_in = nodal_outflow
  []
  
    [outflow_cancel]
    type = PorousFlowOutflowBC
    boundary = 'DRAINSRF_F009_UG5'
    variable = pp
    multiplier = -1
  []
[]
```

Please let me know if it's not right.

Kind regards,
Traiwit

**Reply by GiudGiud** [reply]
The _normals is the normal of the boundary near that element.
so you want to be using the surface velocity rather than the element velocity?
But the velocity is an auxiliary variable right? You only have it computed on a constant per-element basis?

**Reply by Traiwit** [reply]
> Comment options
> ####
> [ GiudGiud ](/GiudGiud) [ May 5, 2023 ](#discussioncomment-5817576)
> Collaborator

Hi @GiudGiud 

I think this is the closest to what we want:

```
AdvectiveFluxAuxBE::computeValue()
{
  // RealVectorValue velocity(_v[_qp]);
  return _vx[_qp] * _normals[_qp](0) + _vy[_qp] * _normals[_qp](1) + (_vz[_qp]) * _normals[_qp](2) ;
}
```

but not sure if it's right comparing to this:

![image](https://user-images.githubusercontent.com/79562439/236589618-188fea7f-fed0-4053-9170-ebf80773dc72.png)

please give me a comment when you have time.

Kind regards,
Traiwit

**Reply by GiudGiud** [reply]
For me, it s just a matter of where things are evaluated, and if you are fine with approximating the velocity with its value at the center of the element or if you want the face value of the velocity. 

If you want the latter, we need to get working on making the porous flow object that can do this (compute a face velocity, or at least not a constant per element velocity we can evaluate on the face)

The _normals vector is the surface normal on this boundary you specified.
